### ISSUE TITLE
libzmq msg type assertion failure when combining zmq and czmq

### ISSUE THREAD

**Issue by kevinkreiser** [zeromq/czmq#1126]
I read @hintjens awesome blog about service discovery and beacons and found that it ended up being implemented in the czmq binding. Since I've been consuming libzmq by wrapping it in c++ classes and throwing it in a library I decided to add libczmq in as well for beacon support. It seems whenever I mix the two I end up with this error as soon as I make ~~(or maybe destroy)~~ a beacon:

    Assertion failed: check () (src/msg.cpp:271)

Which is this: https://github.com/zeromq/libzmq/blob/master/src/msg.cpp#L51

I've worked up a minimal example to show this problem in action. But honestly I don't know enough about the innards of zmq to tell if I'm misusing it or otherwise. Anyway to show the problem all you must do is the following bash script:

    #!/bin/bash
    set -e

    #make the library header file
    cat << EOF > ./doesnt.hpp
    #ifndef __DOESNT__
    #define __DOESNT__
    #include <czmq.h>
    #include <memory>

    struct ctx {
      ctx();
      std::shared_ptr<void> p;
    };

    struct doesnt {
      doesnt();
      std::shared_ptr<zactor_t> p;
    };

    #endif //__DOESNT__
    EOF

    #make the library implementation
    cat << EOF > ./doesnt.cpp
    #include "doesnt.hpp"

    ctx::ctx(){
      auto* context = zmq_ctx_new();  //REMOVE THIS LINE AND IT WORKS
    }

    doesnt::doesnt(){
      p.reset(zactor_new(zbeacon, nullptr), [](zactor_t* a){zactor_destroy(&a);});
    }
    EOF

    #build the library
    g++ doesnt.cpp -std=c++11 -c -fPIC -I. -I/usr/local/include
    g++ doesnt.o -shared -fPIC -L/usr/local/lib -lzmq -lczmq -std=c++11 -Wl,-soname,libdoesnt.so -o libdoesnt.so

    #write a test application that uses the library
    cat << EOF > ./wtf.cpp
    #include <doesnt.hpp>

    int main(void){
      doesnt d;
      return 0;
    }
    EOF

    #run the test application
    g++ wtf.cpp -std=c++11 -I/usr/local/include -I. -L/usr/local/lib -L. -ldoesnt -o wtf
    ./wtf

**Comment by kevinkreiser** [comment]
It probably goes without saying but standard disclaimers about LD_LIBRARY_PATH do apply with respect to running the script above. Make sure you have the locations of both libzmq and czmq in your LD_LIBRARY_PATH etc.

**Comment by kevinkreiser** [comment]
So out of despiration I decided I would build libzmq and libczmq in with debug symbols and without optimization. I did so and magically the above script worked without issue! I reconifgured them and as I normally would, with optimization, and ran it again. It worked again! So.. at this point something was wrong with my install of libzmq or libczmq which caused this but only when used together.. Sorry if anyone spent time looking at this!
